hasktorch/src/Torch/Data/Pipeline.hs

```diff
@@ -11,145 +12,79 @@
-            async $ runEffect $ runTransforms transforms fromTransformBox toBatches
-            pure $ (foldFromProducer (takeBatch fromBatches), batchThread)
-
-makeFold :: (Dataset m dataset batch, MonadIO m)
-    => dataset
-    -> m (L.FoldM m batch b -> m b)
-makeFold = fmap fst . makeFold' 
-
-makeFoldWithTransform :: (MonadIO m, Dataset m dataset batch)  
-  => (batch -> batch')
-  -> dataset
-  -> m (L.FoldM m batch' b -> m b)
-makeFoldWithTransform transf = fmap fst . makeFoldWithTransform' transf 
-
-makeConcurrentFold :: (MonadIO m, ConcurrentDataset m dataset batch')
-  => (batch' -> batch)
-  -> dataset
-  -> Int
-  -> m (L.FoldM m batch b -> m b)
-makeConcurrentFold transforms dataset = fmap fst . makeConcurrentFold' transforms dataset
-  
-foldFromProducer :: Monad m => Producer batch m () -> L.FoldM m batch b -> m b
-foldFromProducer prod fold = (L.impurely P.foldM) fold prod
+import           Control.Monad.Cont (ContT)
+import           Torch.Data.Internal
+import qualified Torch as Torch
+import Lens.Family (view)
+import Pipes.Group
+import Data.Vector (Vector)
+
+data MapStyleOptions = MapStyleOptions { bufferSize :: Int
+                                       , numWorkers :: Int
+                                       }
+
+mapStyleOpts numWorkers = MapStyleOptions { bufferSize = numWorkers
+                                          , numWorkers = numWorkers
+                                          }
+
+class Dataset dataset sample | dataset -> sample  where
+  getItem :: dataset -> Int -> sample
+  size :: dataset -> Int
+
+-- | A sampler takes the range of indexes it needs to produce,
+-- and the size of the dataset it is processing, and returns
+-- indexes to process. The range of indexes is determined by
+-- the number of workers, so with 1 worker the range is the entire dataset,
+-- with 2 each workers range is half the size of the dataset.
+type Sampler m = (Int, Int) -> Int -> Producer Int m ()
+
+-- | A collation pipe that batches a sequence of processed samples.
+type CollateFn sample batch m = Producer sample m () -> Producer batch m ()
+
+makeListT :: forall sampler dataset sample batch m b. (Dataset dataset sample, MonadBaseControl IO m) =>
+  MapStyleOptions -> dataset -> Sampler m -> CollateFn sample batch m -> ContT b m (ListT m (batch, Int))
+makeListT MapStyleOptions{..} dataset sampler collateFn = runWithBuffer bufferSize streamBatches
+  where
+    streamBatches output = forConcurrently_ [0..numWorkers - 1]  (runWorker output)
+    runWorker output = runEffect . yieldItem output
+    -- yieldItem output workerId = sampler (indices workerId) (size dataset) >-> pipeItems  >-> collateFn >->  toOutput' output
+    yieldItem output workerId = collateFn (sampler (indices workerId) (size dataset) >-> pipeItems) >->  toOutput' output
+    pipeItems = for cat (yield . getItem dataset)
+    indices workerId = (workerId * (size dataset `div` numWorkers), (workerId + 1) * (size dataset `div` numWorkers) - 1)
```

Comment:
if possible, I'd like batching and pulling from the dataset to be separated.

Reply:
Frankly, I don't really even know why I approached it like this. It's not how it is done in the streaming interface, and I agree it shouldn't be done here. I think I was trying to make it look like pytorch or something.